src/irmin-pack/inode.ml

```diff
@@ -1687,12 +1687,79 @@ struct
       let of_concrete t = proof_of_concrete (lazy (failwith "blinded root")) t
       let to_concrete = concrete_of_proof ~depth:0
     end
+
+    module Snapshot = struct
+      include T
+
+      type kinded_hash = Contents of hash * metadata | Node of hash
+      [@@deriving irmin]
+
+      type entry = { step : string; hash : kinded_hash } [@@deriving irmin]
+
+      type inode_tree = {
+        depth : int;
+        length : int;
+        pointers : (int * hash) list;
+      }
+      [@@deriving irmin]
+
+      type v = Inode_tree of inode_tree | Inode_value of entry list
+      [@@deriving irmin]
+
+      type inode = { v : v; root : bool } [@@deriving irmin]
+    end
+
+    let get_key ~index h =
+      match index h with
+      | None ->
+          Fmt.failwith
+            "get_key: You are trying to save to the backend an inode \
+             deserialized using [Irmin.Type] that used to contain pointer(s) \
+             to inodes which are unknown to the backend. Hash: %a"
+            pp_hash h
+      | Some key -> key
```

Comment:
I think it would be better for `index` to have the `index:(hash -> key)` signature (instead of `index:(hash -> key option)`). You could then raise the exception from a place that has more context.